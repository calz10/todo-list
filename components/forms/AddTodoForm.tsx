import React from 'react'
import { StyleSheet, View } from 'react-native'
import { TextInput } from '../text-input/TextInput'
import { Button } from '../button/Button'
import DateTimePicker from '@react-native-community/datetimepicker';
import { TodoActionTypes } from '@/redux/todos';
import { StatusPicker } from '../picker/StatusPicker';

type ItemPayload = {
  title: string
  description: string
  dueDate: string
  isAddingTodo?: boolean
  status: TodoActionTypes
  testID? :string
}

type AddTodoFormProps = ItemPayload & {
  onChangeText: (key: keyof ItemPayload, value: string | Date | TodoActionTypes) => void
  onSave: () => void
}

export const AddTodoForm = React.memo(({
  title,
  description,
  dueDate,
  isAddingTodo,
  status,
  testID,
  onChangeText,
  onSave
}: AddTodoFormProps) => {
  return (
    <View style={[styles.container]}>
      <TextInput
        placeholder='Add todo'
        label='Title'
        value={title}
        testID={testID + 'title'}
        onChangeText={(val) => onChangeText('title', val)}
      />
      <TextInput
        placeholder='Add description'
        label='Description'
        multiline
        numberOfLines={4}
        style={styles.textArea}
        value={description}
        testID={testID + 'description'}
        onChangeText={(val) => onChangeText('description', val)}
      />
      <DateTimePicker
        value={dueDate ? new Date(dueDate) : new Date()}
        mode='date'
        display='default'
        testID={testID + 'date'}
        onChange={(evt) => {
          onChangeText('dueDate', new Date(evt.nativeEvent.timestamp).toISOString())
        }}
      />
     {!isAddingTodo &&  <StatusPicker 
        status={status}
        onChange={(val) => onChangeText('status', val)}
        testID={testID + 'date'}
        options={[
          {
            label: 'Pending',
            value: TodoActionTypes.PENDING
          },
          {
            label: 'In Progress',
            value: TodoActionTypes.INPROGRESS
          },
          {
            label: 'Completed',
            value: TodoActionTypes.COMPLETED
          }
        ]}
      />}
      <Button
        title={`${isAddingTodo ? 'Add' : 'Update'} Todo`}
        onPress={onSave}
      />
    </View>
  )
})

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'column',
    rowGap: 16,
    width: '100%',
    paddingHorizontal: 16,
  },
  textArea: {
    height: 150,
    backgroundColor: '#e3e3e4',
    borderRadius: 8,
    padding: 8,
  }
})