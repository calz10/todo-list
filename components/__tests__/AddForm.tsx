import * as React from 'react';
import renderer from 'react-test-renderer';

import { AddTodoForm } from '../forms/AddTodoForm';

it(`renders correctly`, () => {
  const tree = renderer.create(<AddTodoForm dueDate={"1730909577228"} description='test' onChangeText={() => {}} onSave={() => {}} status={1} title='Test Tile'/>).toJSON();

  expect(tree).toMatchSnapshot();
});
