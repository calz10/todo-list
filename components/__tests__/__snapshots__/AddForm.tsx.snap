// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    [
      {
        "display": "flex",
        "flexDirection": "column",
        "paddingHorizontal": 16,
        "rowGap": 16,
        "width": "100%",
      },
    ]
  }
>
  <View
    style={
      {
        "display": "flex",
        "flexDirection": "column",
        "gap": 4,
      }
    }
  >
    <Text>
      Title
    </Text>
    <TextInput
      onChangeText={[Function]}
      placeholder="Add todo"
      style={
        [
          undefined,
          {
            "width": "100%",
          },
        ]
      }
      testID="undefinedtitle"
      value="Test Tile"
    />
  </View>
  <View
    style={
      {
        "display": "flex",
        "flexDirection": "column",
        "gap": 4,
      }
    }
  >
    <Text>
      Description
    </Text>
    <TextInput
      multiline={true}
      numberOfLines={4}
      onChangeText={[Function]}
      placeholder="Add description"
      style={
        [
          {
            "backgroundColor": "#e3e3e4",
            "borderRadius": 8,
            "height": 150,
            "padding": 8,
          },
          {
            "width": "100%",
          },
        ]
      }
      testID="undefineddescription"
      value="test"
    />
  </View>
  <RNDateTimePicker
    date={NaN}
    displayIOS="default"
    enabled={true}
    mode="date"
    onChange={[Function]}
    onPickerDismiss={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="undefineddate"
  />
  <View>
    <RNCPicker
      items={
        [
          {
            "label": "Pending",
            "testID": undefined,
            "textColor": undefined,
            "value": "1",
          },
          {
            "label": "In Progress",
            "testID": undefined,
            "textColor": undefined,
            "value": "2",
          },
          {
            "label": "Completed",
            "testID": undefined,
            "textColor": undefined,
            "value": "3",
          },
        ]
      }
      numberOfLines={1}
      onChange={[Function]}
      selectedIndex={0}
      style={
        [
          {
            "height": 216,
          },
          undefined,
        ]
      }
      testID="undefineddate"
    />
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#277dce",
        "borderRadius": 8,
        "height": 44,
        "justifyContent": "center",
        "opacity": 1,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      Update Todo
    </Text>
  </View>
</View>
`;
